driver/sensor: use lis2dh as common driver among multiple sensors

avisconti opened the pull request:
Share lis2dh driver among few accelerometer sensors that has
same register interface: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12, LSM303AGR.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
#7001

avisconti commented:
@erwango, you may want to take a look as well.

ydamigos reviewed (commented):
Tested it on stm32f3_disco with samples/sensor/lsm303dlhc. It works fine.
Using LIS2DH "name" for all accelerometer sensors that has same register interface seems a little confusing though.

galak commented:
> Tested it on stm32f3_disco with samples/sensor/lsm303dlhc. It works fine.
> Using LIS2DH "name" for all accelerometer sensors that has same register interface seems a little confusing though.

I agree, we should support the same compat's as before, even if there is a single or shared driver.

MaureenHelm commented:
> 2\. We may rename lis2dh driver as st_acc33 and lis2mdl as st_mag40. Not sure is worth doing it.

I don't like this option at all, as expect it will cause a lot of confusion with part numbers. Have a look at what @jfischer-phytec-iot 	did with the fxos8700 driver in #10154. In this case, fxos8700 is the "superset part" which supports 6 axis accel+mag. There are other accel-only parts like mma8451 with a compatible memory map, so he modified the fxos8700 driver to support the additional who-am-i's and removed the standalone mma8451 driver.

avisconti commented:
@MaureenHelm @galak 

> Have a look at what @jfischer-phytec-iot did with the fxos8700 driver in #10154. In this case,
> fxos8700 is the "superset part" which supports 6 axis accel+mag. There are other accel-only 
> parts like mma8451 with a compatible memory map, so he modified the fxos8700 driver to
> support the additional who-am-i's and removed the standalone mma8451 driver.

OK, thx. I took a look to this PR commits as well as the discussion.
It seems to me exactly the same as I did here except for one things which has been highlighted by
@galak in PR #10154:

>	 fxos8700@1d {
>		 compatible = "nxp,fxos8700";
>> We should say:
>> compatible = "nxp,mma8451q", "nxp,fxos8700";

So, if my understanding is correct, the only thing I need to do is to retain the dts bindings for all sensors even if I'm using a single driver. Does it make sense to you?

avisconti committed:
driver/sensor: use lis2dh as common driver among multiple sensors

Share lis2dh driver among few accelerometer sensors that has
same register interface: LIS2DH, LIS3DH, LSM303DLHC, LIS2DH12,
LSM303AGR.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@MaureenHelm @galak 
re-pushed keeping the dts binding for lis2dh/lis3d/lsm303dlhc-accel sensor, but sharing single driver
(i.e. lis2dh). The other 2 drivers have been removed.

avisconti commented:
@MaureenHelm 
re-based, re-tested, re-pushed.
As I said previously the current implementation
  1. keep a single driver (i.e. lis2dh) and removes the other two
   2. keeps ALL 3 bindings (i.e. lis2dh, lis3dh, 303dlhc)
  3. concatenates the compatible strings in the .dts files

MaureenHelm commented on boards/arm/stm32f3_disco/dts_fixup.h:
```diff
@@ -0,0 +1,15 @@
+/* LSM303DLHC Accelerometer sensor is using LIS2DH driver */
+#define DT_LIS2DH_DEV_NAME		DT_ST_STM32_I2C_V2_40005400_ST_LSM303DLHC_ACCEL_19_LABEL
+#define DT_LIS2DH_I2C_ADDRESS		DT_ST_STM32_I2C_V2_40005400_ST_LSM303DLHC_ACCEL_19_BASE_ADDRESS
+#define DT_LIS2DH_I2C_MASTER_DEV_NAME	DT_ST_STM32_I2C_V2_40005400_ST_LSM303DLHC_ACCEL_19_BUS_NAME
```
If you first convert the driver to use DT_<COMPAT>_<INSTANCE> defines as in #12579 , then these fixups can go away.

avisconti replied:
I know Maureen, but I wanted to do it later on.
In case you want it right before then we'll need to wait next week (provided that I will not commit suicide during the weekend...).

I'm already planning with @galak to finish all the pending ST sensor for this kind of change

avisconti replied:
It seems CI is complaining because LIS2Dh name.... So, let's wait next week and I'll incorporate in this PR also the LIS2DH work.

MaureenHelm replied:
> provided that I will not commit suicide during the weekend

Please don't joke about suicide. If you're not joking, please go talk to someone

MaureenHelm reviewed (changes requested):
One requested DT change, but otherwise looks good

avisconti committed:
driver/sensor: lis2dh: align driver to auto-generated dts macros

Use auto-generated device tree macros in LIS2DH driver to avoid
usage of dts.fixup files. The triggered interrupt part has been
slightly hacked to automatically understand whether only int1 is
configured or both int1 and int2.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Re-pushed it with an additional commit, i.e. the re-work of the LIS2DH macros so that the dts_fixup file
could be completely removed. To test it I used the Nucleo F401RE + x_nucleo_iks01a2 and
code in #12398.

ydamigos reviewed (approved):
I still find it a little bit confusing to configure LIS2DH in Kconfig files for other STM32 accelerometers (e.g. LSM303DLHC) but I don't have anything else to suggest, so I approve it.

avisconti commented:
> I still find it a little bit confusing to configure LIS2DH in Kconfig files for other STM32 accelerometers (e.g. LSM303DLHC) but I don't have anything else to suggest, so I approve it.

Yes, I understand. Actually in dts we put both compatible, one for the actual name and one for the
common driver. The config name may be confusing, but it is the way adopted even in linux. I stated
in LIS2DH Kconfig the fact that it supports more sensors:

 menuconfig LIS2DH
       bool "LIS2DH Three Axis Accelerometer"
       depends on (I2C && HAS_DTS_I2C) || (SPI && HAS_DTS_SPI)
       help
          Enable SPI/I2C-based driver for LIS2DH, LIS3DH, LSM303DLHC,
         LIS2DH12, LSM303AGR triaxial accelerometer sensors.

avisconti commented:
@MaureenHelm 
Let me know if it is fine now or not.
